fem/fespace.cpp

```diff
@@ -146,6 +146,31 @@ void FiniteElementSpace::CopyProlongationAndRestriction(
    delete perm_mat_tr;
 }
 
+void FiniteElementSpace::SetProlongation(const SparseMatrix& p)
+{
+   if (!cP)
+   {
+      cP = std::unique_ptr<SparseMatrix>(new SparseMatrix(p));
+   }
+   else
+   {
+      *cP = p;
+   }
+   cP_is_set = true;
+}
+
+void FiniteElementSpace::SetRestriction(const SparseMatrix& r)
+{
+   if (!cR)
+   {
+      cR = std::unique_ptr<SparseMatrix>(new SparseMatrix(r));
+   }
+   else
+   {
+      *cR = r;
+   }
+}
+
```

Comment:
This could be helpful. In `FiniteElementSpace` and `ParFiniteElementSpace`, there is usage of `SparseMatrix`, `HypreParMatrix`, and `Operator` for restriction and prolongation operators. This can be confusing, but on the other hand a user has to know what they are doing when working with such operators. Adding a verification like this could be informative, although I would expect the user to know not to set serial operators for a parallel space.

Reply:
I went ahead an added these checks in the latest commit.  It looks like the `ParFiniteElementSpace` class does use a serial `SparseMatrix` for restriction, but the serial implementation in `FiniteElementSpace` for `SetRestriction` would still be invalid, since the serial class uses the member `FiniteElementSpace::cR`, while the parallel class uses `ParFiniteElementSpace::R`.